File: BrokerClient/ClientPool.cs
using System;
using System.Linq;
using BrokerCommon;
using BrokerCommon.Models;

namespace BrokerClient
{
    public class ClientPool
    {
        internal readonly ClientBrokerManager clientBrokerManager;
        private readonly Func<string, Swimmer> _getSwimmer;
        public string PoolName { get; set; }
        internal OnMessage onMessage { get; set; }
        internal OnMessageWithResponse onMessageWithResponse { get; set; }

        public ClientPool(ClientBrokerManager clientBrokerManager, GetPoolByNameResponse response, Func<string, Swimmer> getSwimmer)
        {
            this.clientBrokerManager = clientBrokerManager;
            _getSwimmer = getSwimmer;
            PoolName = response.PoolName;
        }
        public void OnMessage(OnMessage callback)
        {
            onMessage += callback;
        }

        public void OnMessageWithResponse(OnMessageWithResponse callback)
        {
            onMessageWithResponse += callback;
        }

        public void GetSwimmers(Action<Swimmer[]> callback)
        {
            var query = Query.Build("GetSwimmers", new QueryParam("PoolName", this.PoolName));

            clientBrokerManager.client.SendMessageWithResponse(query, (response) =>
            {
                callback(
                    response.GetJson<GetSwimmerByPoolResponse>()
                        .Swimmers
                        .Select(a => _getSwimmer(a.Id))
                        .ToArray()
                );
            });

        }

        public void JoinPool(Action callback)
        {
            clientBrokerManager.client.SendMessageWithResponse(
                Query.Build("JoinPool", new QueryParam("PoolName", this.PoolName)),
                (response) =>
                {
                    callback();
                }
            );
        }

        public void SendMessage(Query query)
        {
            query.Add("~ToPool~", this.PoolName);
            clientBrokerManager.client.SendMessage(query);
        }

        public void SendAllMessage(Query query)
        {
            query.Add("~ToPoolAll~", PoolName);
            clientBrokerManager.client.SendMessage(query);
        }

        public void SendMessageWithResponse(Query query, Action<Query> callback)
        {
            query.Add("~ToPool~", PoolName);
            clientBrokerManager.client.SendMessageWithResponse(query, callback);
        }

        public void SendAllMessageWithResponse(Query query, Action<Query> callback)
        {
            query.Add("~ToPoolAll~", PoolName);
            clientBrokerManager.client.SendMessageWithResponse(query, callback);
        }
    }
}

File: BrokeClientTS/src/clientPool.ts
import {ClientBrokerManager} from "./clientBrokerManager";
import {Query, QueryParam} from "./common/query";
import {GetPoolByNameResponse} from "./models/GetPoolByNameResponse";
import { Swimmer} from "./swimmer";
import {GetSwimmerByPoolResponse} from "./models/GetSwimmerByPoolResponse";
import {OnMessage, OnMessageWithResponse, SocketLayer} from "./common/socketLayer";

export class ClientPool {
    public clientBrokerManager: ClientBrokerManager;
    public PoolName: string;
    public onMessage: OnMessage[]=[];
    public onMessageWithResponse: OnMessageWithResponse[]=[];
    private _getSwimmer: (swimmerId: string) => Swimmer;

  
    constructor(clientBrokerManager: ClientBrokerManager, response: GetPoolByNameResponse, getSwimmer: (swimmerId: string) => Swimmer) {
        this.clientBrokerManager = clientBrokerManager;
        this.PoolName = response.PoolName;
        this._getSwimmer = getSwimmer;
    }
    public OnMessage(callback: OnMessage): void {
        this.onMessage.push(callback);
    }

    public OnMessageWithResponse(callback: OnMessageWithResponse): void {
        this.onMessageWithResponse.push(callback);
    }



    public GetSwimmers(callback: (_: Swimmer[]) => void): void {
        var query = Query.Build("GetSwimmers", new QueryParam("PoolName", this.PoolName));
        this.clientBrokerManager.client.SendMessageWithResponse(query,
            (response) => {
                callback((response.GetJson<GetSwimmerByPoolResponse>()).Swimmers
                    .map(a => this._getSwimmer(a.Id)));
            });
    }

    public JoinPool(callback: () => void): void {
        this.clientBrokerManager.client.SendMessageWithResponse(
            Query.Build("JoinPool", new QueryParam("PoolName", this.PoolName)),
            (response) => {
                callback();
            });
    }

    public SendMessage(query: Query): void {
        query.Add("~ToPool~", this.PoolName);
        this.clientBrokerManager.client.SendMessage(query);
    }

    public SendAllMessage(query: Query): void {
        query.Add("~ToPoolAll~", this.PoolName);
        this.clientBrokerManager.client.SendMessage(query);
    }

    public SendMessageWithResponse(query: Query, callback: (message: Query) => void): void {
        query.Add("~ToPool~", this.PoolName);
        this.clientBrokerManager.client.SendMessageWithResponse(query,
            (response) => {
                callback(response);
            });
    }

    public SendAllMessageWithResponse(message: Query, callback: (message:Query) => void): void {
        message.Add("~ToPoolAll~", this.PoolName);
        this.clientBrokerManager.client.SendMessageWithResponse(message,
            (response) => {
                callback(response);
            });
    }

    public invokeMessageWithResponse(from: Swimmer, message: Query, respond: (query: Query) => void) {
        for (var i = 0; i < this.onMessageWithResponse.length; i++) {
            this.onMessageWithResponse[i](from,message, respond);
        }
    }

    public invokeMessage(from:Swimmer, message: Query) {
        for (var i = 0; i < this.onMessage.length; i++) {
            this.onMessage[i](from,message);
        }
    }
}